csvcubed/csvcubed/models/codelistconfig/code_list_config.py

```diff
@@ -0,0 +1,196 @@
+    metadata_from_dict,
+)
+
+CODE_LIST_CONFIG_DEFAULT_URL = "https://purl.org/csv-cubed/code-list-config/v1"
+
+
+@dataclass
+class CodeListConfigSort(DataClassBase):
+    """Model for representing the sort object in code list config."""
+
+    by: str
+    method: str
+
+
+@dataclass
+class CodeListConfigConcept(DataClassBase):
+    """Model for representing a code list concept in code list config."""
+
+    label: str
+    notation: str
+    description: Optional[str] = field(default=None)
+    sort_order: Optional[int] = field(default=None)
+    same_as: Optional[str] = field(default=None)
+    children: List["CodeListConfigConcept"] = field(default_factory=list)
+
+
+@dataclass
+class CodeListConfig(DataClassBase):
+    """Model for representing a code list config."""
+
+    sort: Optional[CodeListConfigSort] = field(default=None)
+    concepts: List[CodeListConfigConcept] = field(default_factory=list)
+    schema: str = field(default=CODE_LIST_CONFIG_DEFAULT_URL)
+    # Using CatalogMetadata in the dataclass requires providing default_factory as otherwise, the metadata itself needs to be provided. Since we want have the metadata at initialisation, the default_factory below is defined.
+    metadata: CatalogMetadata = field(
+        default_factory=lambda: CatalogMetadata("Metadata")
+    )
+
+    def __post_init__(self):
+        # Sorting top-level concepts.
+        self.concepts = self._sort_concepts(self.concepts)
+        self._apply_sort_to_child_concepts(self.concepts)
+
+    def _apply_sort_to_child_concepts(self, concepts: List[CodeListConfigConcept]):
+        """
+        Sorting children in each parent concept seperately.
+        """
+        for concept in concepts:
+            if any(concept.children):
+                concept.children = self._sort_concepts(concept.children)
+                self._apply_sort_to_child_concepts(concept.children)
+
+    def _assign_sort_order_to_concepts(
+        self,
+        concepts_without_sort_order: List[CodeListConfigConcept],
+        user_defined_sort_orders: List[int],
+    ) -> List[CodeListConfigConcept]:
+        """Assinging a sort order to concepts without sort order whilst avoiding conflicts with the sort orders already used by the user."""
+
+        sort_order: int = 0
+        concepts: List[CodeListConfigConcept] = []
+        for concept in concepts_without_sort_order:
+            while sort_order in user_defined_sort_orders:
+                sort_order += 1
+            concept.sort_order = sort_order
+            concepts.append(concept)
+            sort_order += 1
+
+        return concepts
+
+    def _sort_concepts(
+        self, concepts: List[CodeListConfigConcept]
+    ) -> List[CodeListConfigConcept]:
+        """
+        Sorting concepts based on the sort object and sort order defined in the code list json.
+        """
+        # Step 1: Identify sort orders defined by the user in code list config json.
+        user_defined_sort_orders: List[int] = [
```

Comment:
Can you use a set instead of a list? It's more performant when you get to large numbers of concepts. I am aware of a significant number of very large code lists.